pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,500 @@
+		return
+	}
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
+	s, err := t.openStream(ctx, info.ID)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to open stream to node", "addr", addr, "node", node, "err", err)
+		return
+	}
+
+	t.addOutboundStream(node, s)
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), node.Pb()))
+}
+
+func (t *Transport) openStream(ctx context.Context, dest peer.ID) (network.Stream, error) {
+	//  We need the simplest retry mechanism due to the fact that the underlying libp2p's NewStream function dials once:
+	// https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358
+	t.logger.Log(logging.LevelDebug, "opening stream to peer", "src", t.ownID, "dst", dest)
+
+	// The implementation is based on the openStream() function from the RemoteTracer:
+	// https://github.com/libp2p/go-libp2p-pubsub/blob/cbb7bfc1f182e0b765d2856f6a0ea73e34d93602/tracer.go#L280
+	var s network.Stream
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, err = t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
+
+		if err == nil {
+			return s, nil
+		}
+
+		if i >= noLoggingErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		}
+
+		delay := time.NewTimer(maxRetryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, err)
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	var err error
+	var nodeErr *nodeInfoError
+	var streamErr *nilStreamError
+
+	s, err := t.getNodeStream(dest)
+	if errors.As(err, &nodeErr) {
+		return err
+	}
+	if errors.As(err, &streamErr) {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		err = w.Flush()
+	}
+	// If we cannot write to the stream, or we don't have a connection to that peer then try to
+	// connect to the peer.
+	if err != nil || len(t.host.Network().ConnsToPeer(s.Conn().RemotePeer())) == 0 {
```

Comment:
We should have a connection here by now, right? You already called `connectToNode` above so this is a sanity-check, but do we really want to try to connect again with the node if the stream failed? Maybe is just better to restart the connection (`Close() + Open()).

Also we should lock the stream while we write on it to prevent parallel writes. Instead of just locking in `getNodeStream` we should have a lock also while we write on it.

Reply:
> You already called connectToNode above

That was called in case we don't have a stream to write in. 
But what if we have already had a stream, we tried to write in it, and get an error?
But I agree we should close the connection!